### PR TITLE
RD-1222 Add basic DeploymentsView widget

### PR DESCRIPTION
**Pull request by Gelio**
This PR adds some basic functionalities for the `DeploymentsView` widget, mainly:
1. The table that fetches deployments data, with correct set of columns
2. Selecting rows (does not influence anything yet)
3. Deployment status column
4. Columns for subenvironments and subservices use icons

The widget is only available in `development` mode, meaning it won't be shown in production, thanks to `webpack.EnvironmentsPlugin`.

I have left multiple TODO items that will be completed in RD-1224. Some require filters, some require the API to work.

This PR relies on https://github.com/cloudify-cosmo/cloudify-ui-components/pull/58 and https://github.com/cloudify-cosmo/cloudify-manager-install/pull/1086 (meaning you probably need one of the latest manager, unless you change the `permission` setting)

Here is how it works at the moment:

https://user-images.githubusercontent.com/889383/109002015-f2325b80-76a5-11eb-94e8-2b111c8635aa.mp4

## To do

- [x] Use `i18n` instead of hardcoding strings
- [x] ~~Add widget permissions to https://docs.google.com/spreadsheets/d/1UqpR2ZjBeWSRtY4BkxWMfyuXLlDinAjgSLUHdIC0guE/edit#gid=534592050 (I asked for access, but have not received it yet)~~ Not necessary, as the spreadsheet is outdated
- [x] Install new ui-components and ui-common
- [x] Remove redundant `eslint-disable` comments for rules disabled in https://github.com/cloudify-cosmo/cloudify-ui-common/pull/50

### CODE REVIEW

File: app/utils/StageAPI.ts [hunk 1]
```diff
@@ -95,22 +104,39 @@ interface CommonWidgetDefinition<Params, Data> {
...
-type RenderCallback<Data, Output> = (widget: Widget, data: Data | null, error: any, toolbox: Toolbox) => Output;
+/**
+ * The empty object is the default value
+ */
+export type WidgetData<D> = D | Record<string, never> | undefined;
+
+export function isEmptyWidgetData(data: unknown): data is Record<string, never> | undefined {
+    return (
+        data === undefined ||
+        (typeof data === 'object' && data !== null && !Array.isArray(data) && Object.keys(data).length === 0)
```
[the remark this answers is not in the document]
Comment (Gelio) [reply app/utils/StageAPI.ts:118]: `Object.keys(null)` results in a `TypeError`, hence the guard for `null` is necessary

image

File: widgets/deploymentsView/src/columns.tsx [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    sortFieldName?: string;
+    width?: string;
+    tooltip?: ReactNode;
+    render(deployment: any): ReactNode;
+}
+
+type WithOptionalProperties<T, OptionalProperties extends keyof T> = Omit<T, OptionalProperties> &
+    Partial<Pick<T, OptionalProperties>>;
+
+const namelessDeploymentsViewColumnDefinitions: Record<
```
[the remark this answers is not in the document]
Comment (Gelio) [reply widgets/deploymentsView/src/columns.tsx]: Good point, I changed the prefix to `partial` :slightly_smiling_face:

File: app/utils/StageAPI.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (kubama) [reply app/utils/StageAPI.ts:118]: Hm... I wasn't aware that `typeof null === 'object'`. In fact I'm surprised 🙂.

File: widgets/deploymentsView/src/renderDeploymentRow.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import { ReactNode } from 'react';
+import type { Toolbox } from '../../../app/utils/StageAPI';
+import { deploymentsViewColumnDefinitions, DeploymentsViewColumnId } from './columns';
+
+const renderDeploymentRow = (toolbox: Toolbox, fieldsToShow: DeploymentsViewColumnId[]) => (deployment: any) => {
+    const { DataTable } = Stage.Basic;
+    const selectedDeploymentId = toolbox.getContext().getValue('deploymentId');
```
[the remark this answers is not in the document]
Comment (Gelio) [reply widgets/deploymentsView/src/renderDeploymentRow.tsx:7]: I don't see why it would not react :sweat_smile: Here are the results:

https://user-images.githubusercontent.com/889383/109140728-33397700-775d-11eb-8226-c47494c571cd.mp4

It's actually the Filter widget that does not respond to `deploymentId` being changed :slightly_smiling_face:

File: app/utils/StageAPI.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gelio) [reply app/utils/StageAPI.ts:118]: Yeah :smile: One of the oldest JS quirks :smile: I have fallen for it too :slightly_smiling_face: Thankfully TS knows `typeof null` and help guard against it :smile:

File: widgets/deploymentsView/src/renderDeploymentRow.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (kubama) [reply widgets/deploymentsView/src/renderDeploymentRow.tsx:7]: Am I missing something? So what's triggering widget rerender on context value change?
I'm asking as I had similar situation in labels widget and it wasn't reacting.

[the remark this answers is not in the document]
Comment (Gelio) [reply widgets/deploymentsView/src/renderDeploymentRow.tsx:7]: Not sure what the case was in the widget you worked on, but as I see it:

`Widget` is connected to the store, reading the `context` (https://github.com/cloudify-cosmo/cloudify-stage/blob/994497a81882bd8512a1c04c7b5379a459417730/app/containers/Widget.js#L13). Thus, any changes to the context will cause every Widget to rerender (https://github.com/cloudify-cosmo/cloudify-stage/blob/609a053231c90785d016d1cebfa06d9323132d34/app/components/Widget.jsx#L222-L230)

Perhaps you were expecting a change in `context` to trigger another `fetchData` call for the widget - that's not true, as I see it here: https://github.com/cloudify-cosmo/cloudify-stage/blob/209e56ed7bde25ef909bfd69ea45e927d461da3e/app/components/WidgetDynamicContent.jsx#L50-L79
It reacts only to changes in the configuration and the tenant :slightly_smiling_face:

File: app/utils/StageAPI.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (kubama) [reply app/utils/StageAPI.ts:118]: If so, I would suggest changing the order of checks - null inequality is the most 'high level' check so it should be done first, it's also less surprising this way:
```suggestion
        (data !== null && typeof data === 'object' && !Array.isArray(data) && Object.keys(data).length === 0)
```

File: widgets/deploymentsView/src/renderDeploymentRow.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (kubama) [reply widgets/deploymentsView/src/renderDeploymentRow.tsx:7]: > Perhaps you were expecting a change in `context` to trigger another `fetchData` call for the widget

Actually this was the case. So it's all clear now. Thanks for a quick response - you surely saved me some time.

File: app/utils/StageAPI.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (kubama) [reply app/utils/StageAPI.ts:118]: You could also use lodash and shorten this expression to three, or maybe even two checks.

File: widgets/deploymentsView/src/renderDeploymentRow.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gelio) [reply widgets/deploymentsView/src/renderDeploymentRow.tsx:7]: No problem :smile: I'm glad to help

File: app/utils/StageAPI.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gelio) [reply app/utils/StageAPI.ts:118]: Sure, I'll adjust it and try to use the (dreaded by myself) lodash :smile:

File: widgets/deploymentsView/src/renderDeploymentRow.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+                <DataTable.Data key={columnId}>{columnDefinition.render(deployment)}</DataTable.Data>
+            ))}
+        </DataTable.Row>,
+        progressUnderline && (
+            <DataTable.Row key={`${deployment.id}-progress`} className="deployment-progress-row">
+                <DataTable.Data className="deployment-progress-row-cell" colSpan={fieldsToShow.length}>
+                    {progressUnderline}
+                </DataTable.Data>
+            </DataTable.Row>
+        )
```
Comment (qooban) [inline widgets/deploymentsView/src/renderDeploymentRow.tsx:27]: Have you tried working with [`Progress` component](https://react.semantic-ui.com/modules/progress)?
I know it can be attached to a `Segment`, not sure how it works with `Table`.

File: widgets/deploymentsView/src/styles.scss [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    'failure': red,
+    'pending': blue
+);
+
+.deploymentsViewWidget {
+    --deployment-progress-underline-height: 0.14em;
+    // NOTE: use CSS custom properties to make user overrides easier
+    @each $stateName, $color in $deploymentStateColors {
+        --deployment-color-#{$stateName}: #{$color};
+    }
```
Comment (qooban) [inline widgets/deploymentsView/src/styles.scss:14]: Nicely done 👍🏻

File: widgets/deploymentsView/src/widget.tsx [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+                default: Object.values(deploymentsViewColumnIds).filter(columnId => columnId !== 'environmentType'),
+                type: Stage.Basic.GenericField.MULTI_SELECT_LIST_TYPE
+            },
+            Stage.GenericConfig.PAGE_SIZE_CONFIG(100),
+            Stage.GenericConfig.SORT_COLUMN_CONFIG('created_at'),
+            Stage.GenericConfig.SORT_ASCENDING_CONFIG(false)
+        ],
+        isReact: true,
+        hasReadme: true,
+        hasStyle: false,
```
[the remark this answers is not in the document]
Comment (qooban) [reply widgets/deploymentsView/src/widget.tsx:71]: Historical reasons, I think. Support for `widget.css` was since I remember.

**Review by qooban — COMMENTED** [review]
Nicely done 👍🏻 

Did not went very deep in review. Did not any UI tests.
TS parts are - at least for me - magical, so it would be difficult for me to spot there any issues at this stage.

**Comment by qooban** [comment]
@Gelio 

Regarding this one:
> Add widget permissions to https://docs.google.com/spreadsheets/d/1UqpR2ZjBeWSRtY4BkxWMfyuXLlDinAjgSLUHdIC0guE/edit#gid=534592050 (I asked for access, but have not received it yet)

I'm not sure it is still maintained. Can you check with Alex and let us know? 
Probably we can remove updating this doc from [Adding new built-in widget to Cloudify UI](https://cloudifysource.atlassian.net/wiki/spaces/RD/pages/435421230/Adding+new+built-in+widget+to+Cloudify+UI) page.

File: app/utils/stageUtils.ts [hunk 7]
```diff
@@ -28,14 +30,12 @@ export default class StageUtils {
 
     static parseMarkdown = marked;
 
-    static saveAs(...args) {
-        saveAs(args);
-    }
+    static saveAs = saveAs;
```
[the remark this answers is not in the document]
Comment (Gelio) [reply app/utils/stageUtils.ts:33]: The code change is the same, so I hope for no conflicts :smile: Let's see how git handles that :crossed_fingers:

File: app/utils/shared/TimeUtils.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
+/**
+ * Created by jakub.niezgoda on 06/02/2019.
+ */
+
```
[the remark this answers is not in the document]
Comment (Gelio) [reply app/utils/shared/TimeUtils.ts:4]: Good point, done :slightly_smiling_face:

File: app/utils/stageUtils.ts [hunk 9]
```diff
@@ -51,31 +51,21 @@ export default class StageUtils {
...
     /**
      * @deprecated use TimeUtils.formatLocalTimestamp
-     * @param timestamp
-     * @param outputPattern
-     * @param inputPattern
      */
-    static formatLocalTimestamp(timestamp, outputPattern = 'DD-MM-YYYY HH:mm', inputPattern = undefined) {
-        return TimeUtils.formatLocalTimestamp(timestamp, outputPattern, inputPattern);
-    }
+    static formatLocalTimestamp = TimeUtils.formatLocalTimestamp.bind(TimeUtils);
```
[the remark this answers is not in the document]
Comment (Gelio) [reply app/utils/stageUtils.ts:60]: Thanks! :smile:

File: widgets/deploymentsView/README.md [hunk 10]
```diff
@@ -0,0 +1,5 @@
+# Deployments View
+
+A complete deployments view – Deployment list, map view, and detailed deployment info
+
+TODO: fill in
```
[the remark this answers is not in the document]
Comment (Gelio) [reply widgets/deploymentsView/README.md:5]: I'll handle that in https://cloudifysource.atlassian.net/browse/RD-1532 then. I'll mention the ticket in the comment

File: widgets/deploymentsView/src/renderDeploymentRow.tsx (hunk 4 above)
Comment (Gelio) [reply widgets/deploymentsView/src/renderDeploymentRow.tsx:27]: Thanks for pointing it out, I'll try that tomorrow :slightly_smiling_face:

File: widgets/deploymentsView/src/styles.scss (hunk 5 above)
Comment (Gelio) [reply widgets/deploymentsView/src/styles.scss:14]: I'm glad you like it :slightly_smiling_face:

**Comment by Gelio** [comment]
@qooban
> I'm not sure it is still maintained. Can you check with Alex and let us know?

 After asking Alex if it should be maintained he said:
> Not that I’m aware of, i things it’s an old one, the best is to ask Jakub n

meaning it's probably outdated and does not require updates. I'll remove it from the instructions in Confluence :slightly_smiling_face:

File: widgets/deploymentsView/src/renderDeploymentRow.tsx (hunk 4 above)
Comment (Gelio) [reply widgets/deploymentsView/src/renderDeploymentRow.tsx:27]: I've tried using the `Progress` component :smile: Here are my findings:

1. Rendering it inside of the table's `tr` is not satisfactory (notice it appeared as the extra column on the right):
    
    image
    
    Here's what I added to `TableRow` https://github.com/cloudify-cosmo/cloudify-ui-components/blob/98dd2d04db6051f12702ded5e378ceb53989e68c/src/components/data/DataTable/TableRow.jsx:
    ```tsx
        <tr
            id={id}
            className={computedClassName}
            onClick={onClick}
            onMouseOver={onMouseOver}
            onMouseOut={onMouseOut}
            onFocus={onMouseOver}
            onBlur={onMouseOut}
            style={_.isFunction(onClick) ? { cursor: 'pointer' } : {}}
        >
            {computedChildren}
            <Progress percent={90} attached="bottom" />
        </tr>
    ```

    Unfortunately, it does not magically attach to the row :slightly_frowning_face: 

    *    btw. due to the logic in https://github.com/cloudify-cosmo/cloudify-ui-components/blob/98dd2d04db6051f12702ded5e378ceb53989e68c/src/components/data/DataTable/TableRow.jsx#L25-L29 it's hard to add a custom `tr` in the table, or a custom `td` for that matter :smile: 
        IMO a bit easier to customize API would be to preserve non-`TableDataCell` elements in the `computedChildren`. That's out of scope for this PR though, I'm just sharing thoughts :slightly_smiling_face: 

2. I could render the `Progress` component instead of doing a custom `div` in `getDeploymentProgressUnderline`
3. Using a `Progress` in the row would make it harder to customize the colors. Overriding CSS variables would not cut it, and the user would have to modify Semantic UI's theme, or change the values in the widget implementation :worried: 

To conclude, using a `Progress` component does not add much value here compared to an easy-enough custom `div`. I recommend staying with the current solution :slightly_smiling_face:

Comment (qooban) [reply widgets/deploymentsView/src/renderDeploymentRow.tsx:27]: Yep, thanks for checking, it's a pity, but definitely make sense not to use it here 👍🏻

File: app/utils/StageAPI.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gelio) [reply app/utils/StageAPI.ts:118]: I gave lodash a shot and I came up with 
```suggestion
        (!!data && !Array.isArray(data) && isEmpty(data))
```


which is a bit of an improvement, but I could also do:

```suggestion
        (!!data&& !Array.isArray(data) && Object.keys(data).length === 0)
```

yet TypeScript is not happy then.

However, I added a few more test cases to see if it still works fine, and unfortunately, the implementation using `lodash` did not pass when calling `isEmptyWidgetData(true)` (it returned `true`)) (expected behavior is `isEmpty(true)`, see [the docs](https://lodash.com/docs/4.17.15#isEmpty)). Thus, even though the implementation may be shorter using lodash, I'd stick to the one that passes all those new tests. The shorter implementation that I pasted above (the one without `typeof`) also does not pass (`Object.keys(true)` is an empty array :disappointed: )

Thanks for letting me experiment and catch those edge cases in unit tests :smile:

[the remark this answers is not in the document]
Comment (kubama) [reply app/utils/StageAPI.ts:118]: Let it be, but I still recommend this: https://github.com/cloudify-cosmo/cloudify-stage/pull/1210/files#r582737173

[the remark this answers is not in the document]
Comment (Gelio) [reply app/utils/StageAPI.ts:118]: Yeah, sorry, I forgot about this one :slightly_smiling_face: 

Funny thing, TypeScript does not like it this way:
image

Looks like it doesn't narrow the types correctly in this case. I reckon it "forgets" about the narrowing of `!=== null` when narrowing with `typeof`, but it works fine the other way around :slightly_smiling_face: 

For more information about narrowing, check out https://www.typescriptlang.org/docs/handbook/2/narrowing.html

**Comment by Gelio** [comment]
I have finally got `cloudify-ui-components` to play nicely with this PR :smile: Aside from upgrading `cloudify-ui-*` packages, I had to upgrade `source-map-loader` due to the following warning during webpack compilation:

```
    WARNING in ./node_modules/react-splitter-layout/lib/index.js
    Module Warning (from ./node_modules/source-map-loader/index.js):
    (Emitted value instead of an instance of Error) Cannot find SourceMap 'data:application/json;charset=utf-8;base64,"+btoa(unescape(encodeURIComponent(JSON.stringify(a))))+"': Error: Can't resolve './data:application/json;charset=utf-8;base64,"+btoa(unescape(encodeURIComponent(JSON.stringify(a))))+"' in '/home/grzegorz/projects/cloudify/cloudify-stage/node_modules/react-splitter-layout/lib'
     @ ./widgets/blueprintSources/src/BlueprintSources.jsx 6:0-51 127:119-133
     @ ./widgets/blueprintSources/src/widget.jsx
```

The warning went away after the upgrade :slightly_smiling_face: 

This makes the PR stable and ready for finalization :slightly_smiling_face:

**Review by kubama — DISMISSED** [review]

**Comment by Gelio** [comment]
docTest failed, as expected, so I raised https://github.com/cloudify-cosmo/docs.getcloudify.org/pull/1725 to update the version of cloudify-ui-components in the docs

I see that there's something weird going on with the bundle size 

```
[2021-02-26T11:39:16.389Z]   Vendor JS bundle

[2021-02-26T11:39:16.389Z]   Package size limit has exceeded by 1.26 MB

[2021-02-26T11:39:16.389Z]   Size limit:   4.1 MB

[2021-02-26T11:39:16.389Z]   Size:         5.36 MB

[2021-02-26T11:39:16.389Z]   Loading time: 109.8 s on slow 3G

[2021-02-26T11:39:16.389Z]   Running time: 1.6 s   on Snapdragon 410

[2021-02-26T11:39:16.389Z]   Total time:   111.3 s
```

I will investigate it on Monday

**Review by qooban — APPROVED** [review]

**Comment by qooban** [comment]
@Gelio 

> I see that there's something weird going on with the bundle size

Just FYI (maybe you are aware of it already) - this page might be useful: https://github.com/cloudify-cosmo/cloudify-stage/blob/master/test/README.md#bundle-size-checks - see last section.

**Comment by Gelio** [comment]
@qooban 

Thanks for the info :smile: 

I tried running `npm run size` some time ago, and it failed. I reran it again, and it failed too :smile: 

image

I reckon I probably need `chromium` or something like that installed locally :slightly_smiling_face: 

I managed to check it anyway using `npm run build:analyse:main`, so I appreciate it was already configured :slightly_smiling_face: The fix is in https://github.com/cloudify-cosmo/cloudify-ui-components/pull/63

**Review by kubama — APPROVED** [review]

**Review by qooban — APPROVED** [review]
